Mathlib/Topology/Order/ScottTopology.lean

```diff
@@ -0,0 +1,439 @@
+@[match_pattern] def toScott : α ≃ WithScottTopology α := Equiv.refl _
+
+/-- `ofScott` is the identity function from the `WithScottTopology` of a type.  -/
+@[match_pattern] def ofScott : WithScottTopology α ≃ α := Equiv.refl _
+
+@[simp] lemma to_scott_symm_eq : (@toScott α).symm = ofScott := rfl
+@[simp] lemma of_scott_symm_eq : (@ofScott α).symm = toScott := rfl
+@[simp] lemma toScott_ofScott (a : WithScottTopology α) : toScott (ofScott a) = a := rfl
+@[simp] lemma ofScott_toScott (a : α) : ofScott (toScott a) = a := rfl
+-- porting note: removed @[simp] to make linter happy
+lemma toScott_inj {a b : α} : toScott a = toScott b ↔ a = b := Iff.rfl
+-- porting note: removed @[simp] to make linter happy
+lemma ofScott_inj {a b : WithScottTopology α} : ofScott a = ofScott b ↔ a = b :=
+Iff.rfl
+
+/-- A recursor for `WithScottTopology`. Use as `induction x using WithScottTopology.rec`. -/
+protected def rec {β : WithScottTopology α → Sort _}
+  (h : ∀ a, β (toScott a)) : ∀ a, β a := fun a => h (ofScott a)
+
+
+instance [Nonempty α] : Nonempty (WithScottTopology α) := ‹Nonempty α›
+instance [Inhabited α] : Inhabited (WithScottTopology α) := ‹Inhabited α›
+
+variable [Preorder α]
+
+instance : Preorder (WithScottTopology α) := ‹Preorder α›
+
+instance : TopologicalSpace (WithScottTopology α) := ScottTopology'
+
+end WithScottTopology
+
+/--
+The Scott topology is defined as the join of the topology of upper sets and the topological space
+where a set `u` is open if, when the least upper bound of a directed set `d` lies in `u` then there
+is a tail of `d` which is a subset of `u`.
+-/
+class ScottTopology (α : Type _) [t : TopologicalSpace α] [Preorder α] : Prop where
+  topology_eq_ScottTopology : t = ScottTopology'
+
+instance [Preorder α] : ScottTopology (WithScottTopology α) :=
+  ⟨rfl⟩
+
+namespace ScottTopology
+
+section preorder
+
+variable [Preorder α]
+
+lemma isOpen_iff_upper_and_Scott_Hausdorff_Open' {u : Set α} :
+  IsOpen (WithScottTopology.ofScott ⁻¹' u) ↔ (IsUpperSet u ∧ ScottHausdorffTopology.IsOpen u) :=
+by rfl
+
+variable [TopologicalSpace α] [ScottTopology α]
+
+variable (α)
+
+lemma topology_eq : ‹_› = ScottTopology' := topology_eq_ScottTopology
+
+variable {α}
+
+/-- If `α` is equipped with the Scott topology, then it is homeomorphic to `WithScottTopology α`.
+-/
+def withScottTopologyHomeomorph : WithScottTopology α ≃ₜ α :=
+  WithScottTopology.ofScott.toHomeomorphOfInducing ⟨by erw [topology_eq α, induced_id]; rfl⟩
+
+end preorder
+
+section morphisms
+
+variable [Preorder α] [Preorder β]
+
+open TopologicalSpace
+
+lemma upperSetTopology_coinduced' {t₁ : TopologicalSpace α} [UpperSetTopology α]
+  (hf : Monotone f) : coinduced f t₁ ≤ @ScottTopology' β _ := by
+  rw [ScottTopology']
+  apply le_sup_of_le_left
+  rwa [← continuous_iff_coinduced_le, ← UpperSetTopology.monotone_iff_continuous]
+
+lemma Monotone_coinduced {t₁ : TopologicalSpace α} [UpperSetTopology α]
```

Comment:
That's a weird lemma name

Reply:
Yes, I still don't have a good feel for multiple topologies on a space. Are these even the right lemmas?